packages/gestalt/src/FormHelperText.js

```diff
@@ -1,21 +1,92 @@
 // @flow strict
-
-import { type Node } from 'react';
+import { Fragment, useRef, useEffect, useState, type Node } from 'react';
 import Box from './Box.js';
+import Flex from './Flex.js';
+import Status from './Status.js';
 import Text from './Text.js';
+import { type MaxLength } from './TextField.js';
 
 type Props = {|
-  text: string,
+  text: ?string,
+  maxLength?: ?MaxLength,
+  currentLength?: number,
   addA11yPause?: boolean,
 |};
 
-export default function FormHelperText({ text, addA11yPause = false }: Props): Node {
+export default function FormHelperText({
+  currentLength,
+  text,
+  maxLength,
+  addA11yPause = false,
+}: Props): Node {
+  const ref = useRef();
+  const [width, setWidth] = useState(undefined);
+
+  useEffect(() => {
+    const containerWidth = ref?.current?.getBoundingClientRect().width;
+    setWidth(containerWidth ? Math.ceil(containerWidth) : undefined);
+  }, [ref]);
+
+  const maxLengthChars = maxLength?.characterCount.toString() ?? '';
+
+  const maxLengthReached = (currentLength ?? 0) >= (maxLength?.characterCount ?? 0);
+
+  let status = 'warning';
+  let textColor = 'warning';
+
+  if (
+    typeof currentLength === 'number' &&
+    typeof maxLength?.characterCount === 'number' &&
+    currentLength > maxLength?.characterCount
+  ) {
+    status = 'problem';
+    textColor = 'error';
+  }
+
   return (
     <Box marginTop={2}>
-      <Text color="subtle" size="100">
-        {addA11yPause ? <Box display="visuallyHidden">:</Box> : null}
-        {text}
-      </Text>
+      <Flex gap={4}>
+        <Flex.Item flex="grow">
+          {text ? (
+            <Text color="subtle" size="100">
+              {addA11yPause ? <Box display="visuallyHidden">:</Box> : null}
+              {text}
+            </Text>
+          ) : null}
+        </Flex.Item>
+        {maxLength ? (
+          <Fragment>
+            {/* This hidden container is used to calculate the width of the character tracker and prevent spacing changes on each input value changes */}
+            <Box
+              position="absolute"
+              dangerouslySetInlineStyle={{ __style: { visibility: 'hidden' } }}
+              ref={ref}
+            >
+              <Text color="subtle" size="100">
+                {`${maxLengthChars}/${maxLengthChars}`}
+              </Text>
+            </Box>
+            <Flex gap={1}>
+              {maxLengthReached ? (
+                <Fragment>
+                  {/* This visually hidden error message is accessible by screenreaders. It alerts the user right after the maximum length is reached. */}
+                  <Box display="visuallyHidden" aria-live="assertive" role="alert">
+                    {maxLength?.errorAccessibilityLabel}
+                  </Box>
+                  <Status type={status} accessibilityLabel={maxLength?.errorAccessibilityLabel} />
```

Comment:
Optional chaining is unnecessary here given the truthy check on line 57

Reply:
<img width="649" alt="Screen Shot 2022-09-22 at 8 20 43 PM" src="https://user-images.githubusercontent.com/10593890/191872607-48656751-0255-4f42-aa6c-bcdddc4e6ff6.png">